File: api/_lib/flex_messages.py
"""LINE Flex Message template builders for vocab cards."""

from __future__ import annotations

from typing import Any

from . import config
from .models import ParsedWord

BRAND_COLOR = config.BRAND_COLOR
MAX_TAGS_PER_CARD = 2
MAX_CAROUSEL_BUBBLES = 10


def build_vocab_card(
    word: ParsedWord,
    card_id: str,
    source_app: str = "General",
) -> dict[str, Any]:
    """Build a single Flex Message bubble for a vocabulary word."""
    # Header: word + pronunciation
    header = {
        "type": "box",
        "layout": "vertical",
        "backgroundColor": BRAND_COLOR,
        "paddingAll": "16px",
        "contents": [
            {
                "type": "text",
                "text": f"📖 {word.word}",
                "color": "#FFFFFF",
                "size": "xl",
                "weight": "bold",
            },
            {
                "type": "text",
                "text": word.pronunciation or " ",
                "color": "#E0FFE0",
                "size": "sm",
                "margin": "xs",
            },
        ],
    }

    # Body: context sentence + translation + tags
    body_contents: list[dict[str, Any]] = []

    if word.context_sentence:
        body_contents.append({
            "type": "text",
            "text": word.context_sentence,
            "size": "md",
            "wrap": True,
            "color": "#333333",
        })

    if word.translation:
        body_contents.append({
            "type": "text",
            "text": f"🇹🇼 {word.translation}",
            "size": "md",
            "wrap": True,
            "color": "#555555",
            "margin": "md",
        })

    if word.context_trans:
        body_contents.append({
            "type": "text",
            "text": word.context_trans,
            "size": "sm",
            "wrap": True,
            "color": "#888888",
            "margin": "sm",
        })

    if word.ai_example:
        body_contents.extend([
            {"type": "separator", "margin": "lg"},
            {
                "type": "text",
                "text": "💡 AI 補充例句",
                "size": "xs",
                "color": "#AAAAAA",
                "margin": "lg",
            },
            {
                "type": "text",
                "text": word.ai_example,
                "size": "sm",
                "wrap": True,
                "color": "#666666",
                "margin": "sm",
            },
        ])

    # Tags row — deduplicate source_app from word.tags
    seen: set[str] = set()
    tag_labels: list[str] = []
    for tag in [source_app] + word.tags[:MAX_TAGS_PER_CARD]:
        if tag and tag not in seen:
            seen.add(tag)
            tag_labels.append(tag)

    tag_contents = []
    for tag in tag_labels:
        tag_contents.append({
            "type": "box",
            "layout": "horizontal",
            "backgroundColor": "#F0F0F0",
            "cornerRadius": "8px",
            "paddingAll": "4px",
            "paddingStart": "8px",
            "paddingEnd": "8px",
            "contents": [
                {
                    "type": "text",
                    "text": f"🏷 {tag}",
                    "size": "xxs",
                    "color": "#888888",
                }
            ],
        })
    if tag_contents:
        body_contents.append({
            "type": "box",
            "layout": "horizontal",
            "spacing": "sm",
            "margin": "lg",
            "contents": tag_contents,
        })

    body = {
        "type": "box",
        "layout": "vertical",
        "paddingAll": "16px",
        "spacing": "sm",
        "contents": body_contents,
    }

    # Footer: action buttons
    footer = {
        "type": "box",
        "layout": "horizontal",
        "spacing": "md",
        "paddingAll": "12px",
        "contents": [
            {
                "type": "button",
                "action": {
                    "type": "postback",
                    "label": "✅ 記住了",
                    "data": f"action=save&card_id={card_id}",
                    "displayText": "✅ 已存入單字本！",
                },
                "style": "primary",
                "color": BRAND_COLOR,
                "height": "sm",
            },
            {
                "type": "button",
                "action": {
                    "type": "postback",
                    "label": "❌ 跳過",
                    "data": f"action=skip&card_id={card_id}",
                    "displayText": "已跳過",
                },
                "style": "secondary",
                "height": "sm",
            },
        ],
    }

    return {
        "type": "bubble",
        "size": "kilo",
        "header": header,
        "body": body,
        "footer": footer,
    }


def build_vocab_carousel(
    words: list[tuple[ParsedWord, str]],
    source_app: str = "General",
) -> dict[str, Any]:
    """
    Build a Flex Message carousel for multiple words.

    Args:
        words: list of (ParsedWord, card_id) tuples
        source_app: detected source application
    """
    bubbles = [
        build_vocab_card(word, card_id, source_app)
        for word, card_id in words[:MAX_CAROUSEL_BUBBLES]
    ]

    if len(bubbles) == 1:
        return {
            "type": "flex",
            "altText": f"📖 單字卡：{words[0][0].word}",
            "contents": bubbles[0],
        }

    return {
        "type": "flex",
        "altText": f"📖 {len(bubbles)} 個單字卡",
        "contents": {
            "type": "carousel",
            "contents": bubbles,
        },
    }


def build_error_message(text: str) -> dict[str, Any]:
    """Build a simple error/info Flex Message."""
    return {
        "type": "flex",
        "altText": text,
        "contents": {
            "type": "bubble",
            "size": "kilo",
            "body": {
                "type": "box",
                "layout": "vertical",
                "paddingAll": "20px",
                "contents": [
                    {
                        "type": "text",
                        "text": "⚠️ SnappWord",
                        "weight": "bold",
                        "size": "md",
                        "color": BRAND_COLOR,
                    },
                    {
                        "type": "text",
                        "text": text,
                        "wrap": True,
                        "size": "sm",
                        "color": "#666666",
                        "margin": "md",
                    },
                ],
            },
        },
    }

File: web/lib/server/flex-messages.ts
/**
 * LINE Flex Message template builders.
 */

import type { ParsedWord } from "./types";

const BRAND_COLOR = "#06C755";
const BRAND_NAME = "SnappWord 截詞";

/** Build a single vocab card bubble. */
function buildVocabCard(
  word: ParsedWord,
  cardId: string,
  sourceApp: string
): Record<string, unknown> {
  const tagBoxes = word.tags.slice(0, 3).map((tag) => ({
    type: "box",
    layout: "horizontal",
    contents: [
      {
        type: "text",
        text: tag,
        size: "xxs",
        color: "#FFFFFF",
        wrap: false,
      },
    ],
    backgroundColor: BRAND_COLOR,
    cornerRadius: "md",
    paddingAll: "4px",
    paddingStart: "8px",
    paddingEnd: "8px",
  }));

  return {
    type: "bubble",
    size: "kilo",
    header: {
      type: "box",
      layout: "horizontal",
      contents: [
        {
          type: "text",
          text: sourceApp,
          size: "xs",
          color: "#FFFFFF",
          weight: "bold",
        },
        {
          type: "text",
          text: word.pronunciation || "",
          size: "xs",
          color: "#FFFFFFB3",
          align: "end",
        },
      ],
      backgroundColor: BRAND_COLOR,
      paddingAll: "12px",
    },
    body: {
      type: "box",
      layout: "vertical",
      contents: [
        {
          type: "text",
          text: word.word,
          size: "xl",
          weight: "bold",
          color: "#2D3436",
          wrap: true,
        },
        {
          type: "text",
          text: word.translation,
          size: "md",
          color: BRAND_COLOR,
          weight: "bold",
          margin: "sm",
        },
        ...(tagBoxes.length > 0
          ? [
              {
                type: "box",
                layout: "horizontal",
                contents: tagBoxes,
                spacing: "xs",
                margin: "md",
              },
            ]
          : []),
        {
          type: "separator",
          margin: "lg",
        },
        ...(word.context_sentence
          ? [
              {
                type: "text",
                text: word.context_sentence,
                size: "sm",
                color: "#636E72",
                wrap: true,
                margin: "lg",
                style: "italic",
              },
              {
                type: "text",
                text: word.context_trans || "",
                size: "xs",
                color: "#636E72",
                wrap: true,
                margin: "sm",
              },
            ]
          : []),
        ...(word.ai_example
          ? [
              {
                type: "text",
                text: `💡 ${word.ai_example}`,
                size: "xs",
                color: "#636E72",
                wrap: true,
                margin: "lg",
              },
            ]
          : []),
      ],
      paddingAll: "16px",
    },
    footer: {
      type: "box",
      layout: "horizontal",
      contents: [
        {
          type: "button",
          action: {
            type: "postback",
            label: "📖 記住了",
            data: `action=save&card_id=${cardId}`,
          },
          style: "primary",
          color: BRAND_COLOR,
          height: "sm",
        },
        {
          type: "button",
          action: {
            type: "postback",
            label: "⏭ 跳過",
            data: `action=skip&card_id=${cardId}`,
          },
          style: "secondary",
          height: "sm",
        },
      ],
      spacing: "sm",
      paddingAll: "12px",
    },
  };
}

/** Build a carousel of vocab cards. */
export function buildVocabCarousel(
  wordCardPairs: [ParsedWord, string][],
  sourceApp: string
): Record<string, unknown> {
  const bubbles = wordCardPairs
    .slice(0, 10)
    .map(([word, cardId]) => buildVocabCard(word, cardId, sourceApp));

  return {
    type: "flex",
    altText: `${BRAND_NAME} - ${wordCardPairs.length} 個新單字`,
    contents: {
      type: "carousel",
      contents: bubbles,
    },
  };
}

/** Build an error/info message. */
export function buildErrorMessage(text: string): Record<string, unknown> {
  return {
    type: "flex",
    altText: text,
    contents: {
      type: "bubble",
      size: "kilo",
      body: {
        type: "box",
        layout: "vertical",
        contents: [
          {
            type: "text",
            text: BRAND_NAME,
            size: "xs",
            color: BRAND_COLOR,
            weight: "bold",
          },
          {
            type: "text",
            text,
            size: "sm",
            color: "#2D3436",
            wrap: true,
            margin: "md",
          },
        ],
        paddingAll: "16px",
      },
    },
  };
}
